turbonomic/kubeturbo
# Support scheduled but unready pods

## Problem

Currently we do not discover pods that are scheduled on a node, but not in ready state. This produces an incorrect state of the cluster:

* Wrong number of pods on a node
* Incorrect resource request usage on a node

The latter causes a more serious problem when market generates wrong Pod move and provision actions.

## Solution

We will discover the pods scheduled on a node but not in ready state. We will set the state of the pod to be:

* Power state: `Unknown`
* Controllable: `false`
* Suspendable: `false`
* Provisonable: `false`
* Monitored: `false`

The resource commodities discovered are `VCPURequest`, and `VMemRequest` only (i.e., no `VCPU` or `VMem` commodities). The quota commodities and access commodities (that lock the pods to node and cluster) are still discovered the same as ready pods. No volume and related commodities are discovered for these pods.

## Implementation
Most of the real changes are in `k8s_discovery_worker.go` and `pod_entity_dto_builder.go`.
* Refactor code to not returning `error` when discovering entities in `k8sDiscoveryWorker`. We already allow discovery to continue with non-fatal errors.
* In `podEntityDTOBuilder`, define resource types to discover for pending pods and running pods, and process pending and running pods separately. Properly set the state/flag for pending pods.
* Update unit tests accordingly.

## Testing
* Check the number of non-ready pods when scoping to a cluster:

![image](https://user-images.githubusercontent.com/10012486/95108005-aa1b9800-0708-11eb-8a23-4af330e11aee.png)

* View the `unknown` pods from the list. We cannot sort pods by state yet:

![image](https://user-images.githubusercontent.com/10012486/95108131-dc2cfa00-0708-11eb-8238-ff7854e98161.png)

* The following is a pod in `Init` status with non-zero resource request:

```
NAME                         READY   STATUS     RESTARTS   AGE
redis-0                      0/2     Init:0/1   0          32d
```
![image](https://user-images.githubusercontent.com/10012486/95105721-91f64980-0705-11eb-921e-d83b055c3189.png)

* The following is a pod in `` status with **no** resource request:

```
NAME                            READY   STATUS             RESTARTS   AGE
cert-manager-7c4fdf69b7-bc67h   0/1     ImagePullBackOff   0          59d
```
![image](https://user-images.githubusercontent.com/10012486/95106144-21036180-0706-11eb-9588-905310077919.png)
